projects/motor_controller_interface/test/test_motor_controller_fsm.c

```diff
@@ -0,0 +1,435 @@
+  TEST_ASSERT_OK(generic_can_register_rx((GenericCan *)&s_can, prv_assert_cruise_forward_state,
+                                         0x00, 0x00, false,
+                                         (void *)&num_cruise_control_forward_messages));
+  while (num_cruise_control_forward_messages[0] + num_cruise_control_forward_messages[1] <
+         TEST_NUM_MESSAGES_RECEIVED) {
+  }
+
+  TEST_ASSERT_TRUE(num_cruise_control_forward_messages[0] > 0);
+  TEST_ASSERT_TRUE(num_cruise_control_forward_messages[1] > 0);
+}
+
+static void prv_transition_to_forward(void) {
+  // Send a Driver Controls CAN message going forward
+  DriverControlsData data = {
+    .throttle = TEST_FORWARD_THROTTLE_VALUE,
+    .direction = DRIVER_CONTROLS_FORWARD,
+    .cruise_control = 0,
+    .brake_state = DRIVER_CONTROLS_BRAKE_DISENGAGED,
+  };
+  fifo_push(&s_fifo, &data);
+  event_raise(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO, 0);
+
+  // Run the FSM
+  Event e = { 0 };
+  StatusCode status = NUM_STATUS_CODES;
+  do {
+    status = event_process(&e);
+    if (e.id == MOTOR_CONTROLLER_INTERFACE_EVENT_CAN_TX ||
+        e.id == MOTOR_CONTROLLER_INTERFACE_EVENT_CAN_RX ||
+        e.id == MOTOR_CONTROLLER_INTERFACE_EVENT_CAN_FAULT) {
+      TEST_ASSERT_TRUE(fsm_process_event(CAN_FSM, &e));
+    } else {
+      TEST_ASSERT_EQUAL(true, motor_controller_fsm_process_event(&e));
+    }
+  } while (status != STATUS_CODE_OK);
+
+  delay_ms(100);
+
+  // Assert that we're still receiving torque mode messages
+  num_forward_messages[0] = 0;
+  num_forward_messages[1] = 0;
+  TEST_ASSERT_OK(generic_can_register_rx((GenericCan *)&s_can, prv_assert_forward_state, 0x00, 0x00,
+                                         false, (void *)&num_forward_messages));
+  while (num_forward_messages[0] + num_forward_messages[1] < TEST_NUM_MESSAGES_RECEIVED) {
+  }
+
+  TEST_ASSERT_TRUE(num_forward_messages[0] > 0);
+  TEST_ASSERT_TRUE(num_forward_messages[1] > 0);
+}
+
+static void prv_transition_to_braking(void) {
+  const CANMessage msg = { 0 };
+
+  // Send a Driver Controls CAN message to enter braking state
+  DriverControlsData data = {
+    .throttle = 0,
+    .direction = DRIVER_CONTROLS_FORWARD,
+    .cruise_control = 0,
+    .brake_state = DRIVER_CONTROLS_BRAKE_ENGAGED,
+  };
+  fifo_push(&s_fifo, &data);
+  event_raise(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO, 0);
+
+  Event e = { 0 };
+  StatusCode status = NUM_STATUS_CODES;
+  do {
+    status = event_process(&e);
+    if (e.id == MOTOR_CONTROLLER_INTERFACE_EVENT_CAN_TX ||
+        e.id == MOTOR_CONTROLLER_INTERFACE_EVENT_CAN_RX ||
+        e.id == MOTOR_CONTROLLER_INTERFACE_EVENT_CAN_FAULT) {
+      TEST_ASSERT_TRUE(fsm_process_event(CAN_FSM, &e));
+    } else {
+      TEST_ASSERT_EQUAL(true, motor_controller_fsm_process_event(&e));
+    }
+  } while (status != STATUS_CODE_OK);
+
+  delay_ms(50);
+  TEST_ASSERT_OK(generic_can_register_rx((GenericCan *)&s_can, prv_assert_braking_state, 0x00, 0x00,
```

Comment:
Seems weird to be registering the RX handlers here. If you call `prv_transition_to_braking` twice, `prv_assert_braking_state` will be called twice for every message.

Reply:
Yeah, I agree! :)

You can probably get around that by passing another parameter to the `transition_` functions to determine whether or not to register additional RX handlers, but I'm not convinced that's a good solution, and it seems a little jank.

The RX handlers seem to be the nicest way to observe the CAN output, which is what we really care about.

Instead of registering them in each transition, maybe it makes more sense to register them globally, and then use flag variables to determine which assertions are currently valid?

I guess what I really need is an API to unregister a `RX` handler, as the only way of clearing handlers is by calling the `init` function? But calling `init` again would override all the CAN storage that the FSM maintains.

Reply:
Since you can filter based on the whole message, you should be able to register RX handlers for messages from your device.

Reply:
Not really, because I need to filter on the message data.